Batteries/Data/BinaryHeap.lean

```diff
@@ -1,78 +1,85 @@
-    (Nat.le_add_right ..)
-  have right_le : i ≤ right := Nat.le_trans left_le (Nat.le_add_right ..)
-  have i_le : i ≤ i := Nat.le_refl _
-  have j : {j : Fin a.size // i ≤ j} := if h : left < a.size then
-    if lt (a.get i) (a.get ⟨left, h⟩) then ⟨⟨left, h⟩, left_le⟩ else ⟨i, i_le⟩ else ⟨i, i_le⟩
-  have j := if h : right < a.size then
-    if lt (a.get j) (a.get ⟨right, h⟩) then ⟨⟨right, h⟩, right_le⟩ else j else j
-  if h : i.1 = j then ⟨a, rfl⟩ else
-    let a' := a.swap i j
-    let j' := ⟨j, by rw [a.size_swap i j]; exact j.1.2⟩
-    have : a'.size - j < a.size - i := by
-      rw [a.size_swap i j]; exact Nat.sub_lt_sub_left i.2 <| Nat.lt_of_le_of_ne j.2 h
-    let ⟨a₂, h₂⟩ := heapifyDown lt a' j'
-    ⟨a₂, h₂.trans (a.size_swap i j)⟩
-termination_by a.size - i
-
-@[simp] theorem size_heapifyDown (lt : α → α → Bool) (a : Array α) (i : Fin a.size) :
-    (heapifyDown lt a i).1.size = a.size := (heapifyDown lt a i).2
+  if hleft : left < sz then
+    if hright : right < sz then
+      if lt a[left] a[right] then
+        some ⟨right, hright⟩
+      else
+        some ⟨left, hleft⟩
+    else
+      some ⟨left, hleft⟩
+  else none
+
+/-- Core operation for binary heaps, expressed directly on arrays.
+Given an array which is a max-heap, push item `i` down to restore the max-heap property. -/
+def heapifyDown (lt : α → α → Bool) (a : Vector α sz) (i : Fin sz) :
+    Vector α sz :=
+  match h : maxChild lt a i with
+  | none => a
+  | some j =>
+    have : i < j := by
+      cases i; cases j
+      simp only [maxChild] at h
+      split at h
+      · split at h
+        · split at h <;> (cases h; simp_arith)
+        · cases h; simp_arith
+      · contradiction
+    if lt a[i] a[j] then
+      heapifyDown lt (a.swap i j) j
+    else a
+termination_by sz - i
 
 /-- Core operation for binary heaps, expressed directly on arrays.
 Construct a heap from an unsorted array, by heapifying all the elements. -/
-def mkHeap (lt : α → α → Bool) (a : Array α) : {a' : Array α // a'.size = a.size} :=
+def mkHeap (lt : α → α → Bool) (a : Vector α sz) : Vector α sz :=
   loop (a.size / 2) a (Nat.div_le_self ..)
 where
   /-- Inner loop for `mkHeap`. -/
-  loop : (i : Nat) → (a : Array α) → i ≤ a.size → {a' : Array α // a'.size = a.size}
-  | 0, a, _ => ⟨a, rfl⟩
+  loop : (i : Nat) → (a : Vector α sz) → i ≤ sz → Vector α sz
+  | 0, a, _ => a
   | i+1, a, h =>
-    let h := Nat.lt_of_succ_le h
-    let a' := heapifyDown lt a ⟨i, h⟩
-    let ⟨a₂, h₂⟩ := loop i a' ((heapifyDown ..).2.symm ▸ Nat.le_of_lt h)
-    ⟨a₂, h₂.trans a'.2⟩
+    let a' := heapifyDown lt a ⟨i, Nat.lt_of_succ_le h⟩
+    loop i a' (Nat.le_trans (Nat.le_succ _) h)
 
-@[simp] theorem size_mkHeap (lt : α → α → Bool) (a : Array α) :
-    (mkHeap lt a).1.size = a.size := (mkHeap lt a).2
+private def parent : Fin sz → Option (Fin sz)
+  | ⟨0, _⟩ => none
+  | ⟨i+1, hi⟩ => some ⟨i / 2, Nat.lt_of_le_of_lt (Nat.div_le_self _ _) (Nat.lt_of_succ_lt hi)⟩
```

Comment:
IMO this should just be inlined into `heapifyUp`, saving a branch and also simplifying the code below AFAICT